internal/cluster/node_pool.go

```diff
@@ -45,33 +123,97 @@ type NodePoolStore interface {
 	DeleteNodePool(ctx context.Context, clusterID uint, name string) error
 }
 
+// NodePoolValidator validates a new node pool descriptor.
+type NodePoolValidator interface {
+	// Validate validates a new node pool descriptor.
+	Validate(ctx context.Context, cluster Cluster, rawNodePool NewRawNodePool) error
+}
+
 // NodePoolManager manages node pool infrastructure.
 type NodePoolManager interface {
+	// CreateNodePool creates a new node pool in a cluster.
+	CreateNodePool(ctx context.Context, clusterID uint, rawNodePool NewRawNodePool) error
+
 	// DeleteNodePool deletes a node pool from a cluster.
 	DeleteNodePool(ctx context.Context, clusterID uint, name string) error
 }
 
 // NewNodePoolService returns a new NodePoolService.
-func NewNodePoolService(clusters Store, nodePools NodePoolStore, manager NodePoolManager) NodePoolService {
+func NewNodePoolService(
+	clusters Store,
+	nodePools NodePoolStore,
+	validator NodePoolValidator,
+	manager NodePoolManager,
+) NodePoolService {
 	return nodePoolService{
 		clusters:  clusters,
 		nodePools: nodePools,
+		validator: validator,
 		manager:   manager,
 	}
 }
 
-func (s nodePoolService) DeleteNodePool(ctx context.Context, clusterID uint, name string) (bool, error) {
+func (s nodePoolService) CreateNodePool(
+	ctx context.Context,
+	clusterID uint,
+	rawNodePool NewRawNodePool,
+) error {
 	cluster, err := s.clusters.GetCluster(ctx, clusterID)
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	if err := s.supported(cluster); err != nil {
+	if err := s.checkCluster(cluster); err != nil {
+		return err
+	}
+
+	var nodePool NewNodePool
+
+	if err := mapstructure.Decode(rawNodePool, &nodePool); err != nil {
+		return errors.Wrap(err, "failed to decode node pool")
+	}
+
+	if err := nodePool.Validate(); err != nil {
+		return err
+	}
+
+	if err := s.validator.Validate(ctx, cluster, rawNodePool); err != nil {
+		return err
+	}
+
+	exists, err := s.nodePools.NodePoolExists(ctx, clusterID, nodePool.Name)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		return errors.WithStack(NodePoolAlreadyExistsError{
+			ClusterID: clusterID,
+			NodePool:  nodePool.Name,
+		})
+	}
+
+	err = s.clusters.SetStatus(ctx, clusterID, Updating, "creating node pool")
```

Comment:
I wonder if this should be invoked right after  `s.checkCluster(cluster)` or just before `s.nodePools.NodePoolExists` to ensure there won't be two create node pool flows running in parallel in case user invokes `add node pool` API for the same node pool  within a very short period of time.

Reply:
There is no atomicity guaranteed, so even then two *very* close calls could end up adding a new node pool.

Status update should only happen once all checks are passed, otherwise we would have to change the status back. If we want additional guarantees, we should introduce a locking mechanism. But as a first version this should be fine, we don't have such guarantees anywhere else in the system.